lib/codegen/ProofEvent.cpp

```diff
@@ -0,0 +1,198 @@
+  // the MergeBlock.
+  // If the MergeBlock has no instructions, this has resulted in a segfault when
+  // printing the IR. Adding an effective nop prevents this.
+  llvm::BinaryOperator::Create(
+      llvm::Instruction::Add,
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      "nop", MergeBlock);
+
+  return CurrentBlock = MergeBlock;
+}
+
+llvm::BasicBlock *
+ProofEvent::hookEvent_post(llvm::Value *val, KORECompositeSort *sort) {
+  auto b = proofBranch("hookpost");
+  auto TrueBlock = b.first;
+  auto MergeBlock = b.second;
+  auto OutputFileName = Module->getOrInsertGlobal(
+      "output_file", llvm::Type::getInt8PtrTy(Module->getContext()));
+  auto ir = new llvm::IRBuilder(TrueBlock);
+  auto outputFile = new llvm::LoadInst(
+      llvm::Type::getInt8PtrTy(Module->getContext()), OutputFileName, "output",
+      TrueBlock);
+
+  auto cat = sort->getCategory(Definition);
+  std::ostringstream Out;
+  sort->print(Out);
+  auto sortptr = ir->CreateGlobalStringPtr(Out.str(), "", 0, Module);
+
+  writeUInt64(outputFile, Module, 0xbbbbbbbbbbbbbbbb, TrueBlock);
+  if (cat.cat == SortCategory::Symbol || cat.cat == SortCategory::Variable) {
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            getValueType({SortCategory::Symbol, 0}, Module),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  } else if (val->getType()->isIntegerTy()) {
+    val = ir->CreateIntToPtr(
+        val, llvm::Type::getInt8PtrTy(Module->getContext()));
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeRawTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  } else {
+    val = ir->CreatePointerCast(
+        val, llvm::Type::getInt8PtrTy(Module->getContext()));
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeRawTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  }
+
+  llvm::BranchInst::Create(MergeBlock, TrueBlock);
+
+  llvm::BinaryOperator::Create(
+      llvm::Instruction::Add,
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      "nop", MergeBlock);
```

Comment:
This is commented elsewhere, but the change is repeated a couple of times so perhaps it makes sense to extract the code. We could also audit callers to see if the "insert after last instruction" idiom could be changed; you will know better than me how viable that is.

Reply:
Yeah, this can be extracted.

"insert after last instruction" is kind of the standard way of building up instructions in an `llvm::BasicBlock`, so I don't think we want to change that. In this case I think we may have hit a bug in llvm, I'll try opening and issue for it on their project.